lib/sycamore/sycamore/writers/weaviate_writer.py

```diff
@@ -1,145 +1,235 @@
-    def __init__(
-        self,
-        plan: Node,
-        collection_name: str,
-        client_params: dict,
-        collection_config: Optional[dict],
-        **ray_remote_args,
-    ):
-        super().__init__(plan, **ray_remote_args)
-        self.collection_name = collection_name
-        self.client_params = client_params
-        self.collection_config = collection_config
-
-    def execute(self) -> Dataset:
-        dataset = self.child().execute()
-        datasink = WeaviateDatasink(self.collection_name, self.client_params, self.collection_config)
-        dataset.write_datasink(datasink, ray_remote_args=self.resource_args)
-        return dataset
-
-
-class WeaviateDatasink(Datasink):
-    def __init__(self, collection_name: str, client_params: dict, collection_config: Optional[dict]):
-        self._collection_name = collection_name
-        self._client_params = client_params
-        self._collection_config = collection_config
-
-    def on_write_start(self):
-        from weaviate import WeaviateClient
-
-        with WeaviateClient(**self._client_params) as client:
-            if self._collection_config is not None:
-                if client.collections.exists(self._collection_name):
-                    logging.warning(
-                        f"Collection config was provided, but collection {self._collection_name} "
-                        "already exists, so ignoring provided config."
-                    )
+from dataclasses import dataclass, asdict
+from typing import Optional, Union, Any
+from sycamore.writers.common import drop_types, flatten_data
+from typing_extensions import TypeGuard, TypeAlias
+
+from sycamore.data.document import Document
+from sycamore.writers.base import BaseDBWriter
+from weaviate.classes.config import DataType, ReferenceProperty
+from weaviate.client import (
+    AdditionalConfig,
+    AuthCredentials,
+    ConnectionParams,
+    EmbeddedOptions,
+    UnexpectedStatusCodeError,
+)
+from weaviate.client import WeaviateClient as WvC
+from weaviate.collections.classes.config import (
+    _CollectionConfigCreate,
+    CollectionConfig,
+)
+from weaviate.util import WeaviateInvalidInputError
+
+
+@dataclass
+class WeaviateClientParams(BaseDBWriter.ClientParams):
+    connection_params: Optional[ConnectionParams] = None
+    embedded_options: Optional[EmbeddedOptions] = None
+    auth_client_secret: Optional[AuthCredentials] = None
+    additional_headers: Optional[dict] = None
+    additional_config: Optional[AdditionalConfig] = None
+    skip_init_checks: bool = False
+
+
+# This is mainsly so people don't feel weird about importing
+# this class. This kinda stuff is all over the wv8 codebase
+CollectionConfigCreate: TypeAlias = _CollectionConfigCreate
+
+
+@dataclass
+class WeaviateTargetParams(BaseDBWriter.TargetParams):
+    name: str
+    collection_config: Union[CollectionConfigCreate, CollectionConfig]
+
+    def compatible_with(self, other: BaseDBWriter.TargetParams) -> bool:
```

Comment:
This may just be my ignorance on the base model, but when is this called?

Reply:
[here](https://github.com/aryn-ai/sycamore/blob/87d3123ae23c686512b010f18ccaf96d9c5258cf/lib/sycamore/sycamore/writers/base.py#L71). The target creation logic is:
1. Try to create the target, but don't fail if it's already there
2. Get metadata about the target and verify that it matches what you expect
3. Write documents

This lets us scale writes horizontally without worrying about a singular 'start_write' step, and also lets us catch trying to write to a target that exists already and is incompatible (e.g. diff vector dimensions or something)